### PR TITLE
Use rsp files when linking many files with gnulike linker

### LINKED ISSUE #7212: Response files needed sometimes even on non-windows systems
A user porting www.flintlib.org to meson ran into a commandline length limit
when it tried to link > 3300 files into a single library, with error
```
ninja: fatal: posix_spawn: Argument list too long
```
https://www.in-ulm.de/~mascheck/various/argmax/ 
says cygwin has the tightest limit on commandline length, but FreeBSD also has a smallish limit.
I was able to reproduce the error on FreeBSD with the following script:
```
#!/bin/sh
set -ex
n=1
echo "project('toolong', 'c')" > meson.build
echo "srcs = [" >> meson.build
while test $n -lt 12000
do
    echo "int f$n(int x) { return x + $n; }" > foo$n.c
    echo " 'foo$n.c'," >> meson.build
    n=$(expr $n + 1)
done
echo "]" >> meson.build
echo "library('toolong', srcs)" >> meson.build
meson . btmp
cd btmp
ninja -v
```

CMake successfully avoids the problem by using a response file, and meson should, too.
(Though ideally it should avoid them when not needed, as they do make debugging a little harder.)

### PR DESCRIPTION
**Pull request by dankegel**
Even non-windows systems have commandline limits, so when linking
"too many" files, use a response file.  (Don't use them on normal-sized
link commands; they're obscure and would make developers unhappy.)

This patch uses a lame definition of "too many", and doesn't address static libraries,
but it may be enough for now.

Partially fixes #7212

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -2595,7 +2602,15 @@ def generate_link(self, target, outname, obj_list, linker, extra_args=None, stdl
...
-        elem = NinjaBuildElement(self.all_outputs, outname, linker_rule, obj_list, implicit_outs=implicit_outs)
+        # On Windows systems, always use rsp files.
+        # On other systems, only use them for long lines passed to dynamic linker (for now)
+        # FIXME: don't hardcode commandline length limit; do define it better
+        max_len = 1000
+        if linker_base == 'STATIC' or mesonlib.is_windows() or len(obj_list) > max_len:
+            rule = linker_rule
+        else:
+            rule = linker_short_rule
+        elem = NinjaBuildElement(self.all_outputs, outname, rule, obj_list, implicit_outs=implicit_outs)
```
Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py:2613]: Extract the decision logic and its corresponding comment to its own method called `should_use_rsp_file` or something like that.

**Comment by marc-h38** [comment]
More background info at https://groups.google.com/d/msg/mesonbuild/419IYNF0PAk/m7teKUbqAAAJ

**Comment by jon-turney** [comment]
You might find the test case added in #5245 a useful basis for adding a test for this.

I suspect this won't work in some edge cases, as the quoting rules for response files aren't quite the same as shell, and meson has a pile of Windows-specific hacks to attempt to deal with that.

**Comment by dankegel** [comment]
Thank you.  I did a cursory search for test cases, but missed that one.

**Comment by jon-turney** [comment]
> a useful basis for adding a test for this.

Just to be clear, by that I meant this commit: https://github.com/mesonbuild/meson/pull/5245/commits/e69382a48a2d40ee07c366f11b29262d9f12a057. Sorry for any miscommunication.

**Comment by dankegel** [comment]
Closing in favor of https://github.com/mesonbuild/meson/pull/7227
